nats/bench_command.go

```diff
@@ -147,30 +282,68 @@ func (c *benchCmd) runPublisher(bm *bench.Benchmark, nc *nats.Conn, startwg *syn
 
 	start := time.Now()
 
-	var m *nats.Msg
-	var err error
-
-	errBytes := []byte("error")
-	minusByte := byte('-')
-
-	for i := 0; i < numMsg; i++ {
-		if progress != nil {
-			progress.Incr()
-		}
-
-		if !c.ack {
-			nc.Publish(c.subject, msg)
-			continue
+	if !c.js {
+
+		var m *nats.Msg
+		var err error
+
+		errBytes := []byte("error")
+		minusByte := byte('-')
+
+		for i := 0; i < numMsg; i++ {
+			if progress != nil {
+				progress.Incr()
+			}
+
+			if !c.request {
+				err = nc.Publish(c.subject, msg)
+				if err != nil {
+					log.Fatalf("Publish error: %s", err)
+				}
+			} else {
+				m, err = nc.Request(c.subject, msg, time.Second)
+				if err != nil {
+					log.Fatalf("Request error: %s", err)
+				}
+
+				if len(m.Data) == 0 || m.Data[0] == minusByte || bytes.Contains(m.Data, errBytes) {
+					log.Fatalf("Publish Request did not receive a positive ACK: %q", m.Data)
+				}
+			}
 		}
-
-		m, err = nc.Request(c.subject, msg, time.Second)
+	} else {
+		js, err := nc.JetStream()
 		if err != nil {
-			log.Println(err)
-			continue
+			log.Fatalf("Couldn't create jetstream context: %v", err)
 		}
 
-		if len(m.Data) == 0 || m.Data[0] == minusByte || bytes.Contains(m.Data, errBytes) {
-			log.Printf("Did not receive a positive ACK: %q", m.Data)
+		if !c.syncPub {
+			for i := 0; i < numMsg; i += c.pubBatch {
+				for j := 0; j < c.pubBatch && i+j < c.numMsg; j++ {
+					if progress != nil {
+						progress.Incr()
+					}
+					js.PublishAsync(c.subject, msg)
+					if err != nil {
+						log.Fatalf("PubAsync error: %v", err)
+					}
+				}
+				select {
+				case <-js.PublishAsyncComplete():
+				case <-time.After(time.Second):
+					log.Fatalf("JS PubAsync did not receive a positive ack")
+				}
+			}
+		} else {
+			for i := 0; i < numMsg; i++ {
+				if progress != nil {
+					progress.Incr()
+				}
+				_, err = js.Publish(c.subject, msg)
+				if err != nil {
+					log.Fatalf("Publish error: %s", err)
+				}
+			}
```

Comment:
this method has turned into a bit of a monster with several sets of if js / else branching, might be easier to grok if there was `benchmarkJS` and `benchmarkCore` or similar functions

Reply:
sure...